flink-ml-lib/src/test/java/org/apache/flink/ml/classification/knn/KnnTest.java

```diff
@@ -0,0 +1,287 @@
+    }
+
+    @Test
+    public void testFeaturePredictionParam() throws Exception {
+        Knn knn =
+                new Knn()
+                        .setLabelCol("label")
+                        .setFeaturesCol("features")
+                        .setK(4)
+                        .setPredictionCol("prediction");
+        KnnModel model = knn.fit(trainData);
+        Table output = model.transform(testData)[0];
+
+        assertEquals(
+                Arrays.asList("label", "features", "prediction"),
+                output.getResolvedSchema().getColumnNames());
+
+        List<Tuple2<Double, Double>> result = executeAndCollect(output);
+        verifyClusteringResult(result);
+    }
+
+    @Test
+    public void testFewerDistinctPointsThanCluster() throws Exception {
+        Knn knn =
+                new Knn()
+                        .setLabelCol("label")
+                        .setFeaturesCol("features")
+                        .setK(4)
+                        .setPredictionCol("prediction");
+        KnnModel model = knn.fit(testData);
+        Table output = model.transform(testData)[0];
+
+        assertEquals(
+                Arrays.asList("label", "features", "prediction"),
+                output.getResolvedSchema().getColumnNames());
+        executeAndCollect(output);
+    }
+
+    @Test
+    public void testFitAndPredict() throws Exception {
+        Knn knn =
+                new Knn()
+                        .setLabelCol("label")
+                        .setFeaturesCol("features")
+                        .setK(4)
+                        .setPredictionCol("prediction");
+        KnnModel knnModel = knn.fit(trainData);
+        Table output = knnModel.transform(testData)[0];
+        List<Tuple2<Double, Double>> result = executeAndCollect(output);
+        verifyClusteringResult(result);
+    }
+
+    @Test
+    public void testSaveLoadAndPredict() throws Exception {
+        String path = Files.createTempDirectory("").toString();
+        Knn knn =
+                new Knn()
+                        .setLabelCol("label")
+                        .setFeaturesCol("features")
+                        .setK(4)
+                        .setPredictionCol("prediction");
+        knn.save(path);
+
+        Knn loadKnn = Knn.load(env, path);
+        KnnModel knnModel = loadKnn.fit(trainData);
+        Table output = knnModel.transform(testData)[0];
+
+        List<Tuple2<Double, Double>> result = executeAndCollect(output);
+        verifyClusteringResult(result);
+    }
+
+    @Test
+    public void testModelSaveLoadAndPredict() throws Exception {
+        String path = Files.createTempDirectory("").toString();
+        Knn knn =
+                new Knn()
+                        .setLabelCol("label")
+                        .setFeaturesCol("features")
+                        .setK(4)
+                        .setPredictionCol("prediction");
+        KnnModel knnModel = knn.fit(trainData);
+        knnModel.save(path);
```

Comment:
Could we update the code to use `StageTestUtils.saveAndReload(...)`? 

This util method is provided in the NaiveBayes PR. The main benefit of this approach is to use `org.junit.rules.TemporaryFolder` to make sure the temporary directory will be deleted after the test finishes.

And do you think it would be a bit better to merge `testModelSaveLoadAndPredict` into `testSaveLoadAndPredict`, similar to `NaiveBayesTest::testSaveLoadAndPredict()`? Ideally we could find out the best practice for testing similar functionalities across algorithms so that our algorithm test code will be neat and consistent.

Reply:
OK